bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java

```diff
@@ -360,10 +362,10 @@ public void checkpoint(Checkpoint checkpoint) throws IOException {
 
     @Override
     public synchronized void flush() throws IOException {
-        if (!somethingWritten) {
+        if (!somethingWritten.get()) {
```

Comment:
Use a compareAndSet here. 

Do we need synchronization on the flush method at all?

Reply:
changing it to compareAndSet.

but yeah I considered removing synchronized for flush method, but it is not clear regarding the transactional guarantees of the underlying methods (methods which get called by ILS.flush). So I didn't change this, moreover it is called by requestFlush, which is called by SyncThread.shutdown and SyncThread.start, which are one time events and not related to this work item. So left unchanged.

Reply:
with the compareAndSet, the only thing protected here is the flushOrCheckpoint(), which is called unprotected from checkpoint() above, so I think it would be safe. 

It only seems to be called from the syncthread, which means it's effectively synchronized anyhow since it's all on a single thread, and the synchronized here was only to protect the somethingWritten flag. 

It's no harm to leave it in, just seems weird to be mixing atomics and synchronization without good reason.